WIP added pack manifest cli

ep0ll opened the pull request:
### CNB Pack Manifest Commands: Streamlining Multi-Architecture Image Management
This proposal introduces a new **experimental** set of commands to the CNB tooling ecosystem: _**pack manifest**_. These commands empower buildpack authors to efficiently manage and distribute multi-architecture images, aligning with functionalities offered by existing tools like `podman manifest` and `docker manifest`.

### Key Features:

- **Create manifest lists**: Construct local manifest lists for annotating and pushing to container registries.
- **Annotate manifests**: Add essential information like architecture, OS, or variant to existing manifest lists.
- **Add/remove images**: Manage individual image entries within manifest lists.
- **Push manifests**: Publish constructed manifest lists to container registries.
- **Inspect manifests**: View detailed information about locally stored manifest lists.
### Benefits:

**Simplified workflows**: Centralized management of multi-architecture image variants for efficient distribution.
Enhanced interoperability: Potential adherence to established standards for seamless integration with existing tools and runtimes.
**Improved developer experience**: Streamlined workflows and user-friendly commands for managing manifest lists.
**Target Users**:

Buildpack authors seeking to distribute their buildpacks across diverse architectures.
### Example Workflow:

**Create a manifest list**:

```sh
pack manifest create cnbs/sample-package:hello-multiarch-universe \
    cnbs/sample-package:hello-universe \
    cnbs/sample-package:hello-universe-windows
```
**Annotate architecture**:

```sh
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe-windows
```
**Push to registry**:

```sh
pack manifest push cnbs/sample-package:hello-multiarch-universe
```
**Current Status**:

These commands are initially **experimental** and gated behind the pack config experimental flag. They will transition to supported status upon deemed readiness by maintainers.

This innovative addition to the CNB ecosystem empowers buildpack authors to streamline multi-architecture image management, fostering efficient distribution and enhanced developer experiences.

### Development setup:
This guide outlines the steps to set up your development environment and test the pack manifest commands under development in the image-index branch of the `github.com/WYGIN/buildpacks-pack` repository against the `github.com/WYGIN/buildpacks-imgutil` repository.

#### Prerequisites:
- Docker installed and running on your system.
- Git version control system installed.

#### Steps:

1. **Clone the repositories**:

```sh
git clone https://github.com/WYGIN/buildpacks-pack.git
```

2. **Navigate to the buildpacks-pack repository**:
```sh
cd buildpacks-pack
```

3. **Checkout the `image-index` branch**:
```sh
git checkout image-index
```

4. **Build the pack binary**:
```sh
make build
```

5. **Install the pack binary**:
```sh
sudo make install
```

6. **Enable pack's `experimental features`**:
```sh
pack config experimental true
```

7. **Run `pack manifest`**:

```sh
pack manifest create cnbs/sample-package:hello-multiarch-universe \
     cnbs/sample-package:hello-universe \

     cnbs/sample-package:hello-universe-windows
```

```sh
# annotate first image
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe
# annotate second image
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe-windows
```

```sh
pack manifest add cnbs/sample-package:hello-multiarch-universe \
					cnbs/sample-package:hello-universe-riscv-linux
```

```sh
pack manifest delete cnbs/sample-package:hello-multiarch-universe
```

```sh
pack manifest rm cnbs/sample-package:hello-multiarch-universe \
					cnbs/sample-package:hello-universe-windows
```

```sh
pack manifest push cnbs/sample-package:hello-multiarch-universe
```

```sh
pack manifest inspect cnbs/sample-builder:multiarch
```

#### **Switching between modes**:
There are two modes, where one only manipulates the ImageIndex, where as the other will manipulates the ImageIndex along with the underlying image.
To switch between these modes use:
```sh
pack config index-full-mode [<true> | <false>]
```

ep0ll committed:
WIP added pack manifest cli

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP updated code as per imgutil's ImageIndex PR

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP upgraded imgutil to latest version

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix some bugs of pack manifest cli

Signed-off-by: WYGIN <wygininc@gmail.com>

Commit:
WIP fix manifest cli

Signed-off-by: sai kiran <wyginc1@gmail.com>

ep0ll committed:
WIP fix bugs related to OSFeatures and Push index

Signed-off-by: WYGIN <wygininc@gmail.com>

jjbustamante commented:
Hi @WYGIN,

Thanks a lot for working on this, it's awesome.

I compiled the binaries, and I will try to play with them, first thing I noticed, when I run `pack manifest --help` the short description for each command is been shown.
 
<img width="1682" alt="Screenshot 2024-02-09 at 9 21 17 AM" src="https://github.com/husni-faiz/pack/assets/1181799/f44389ca-4c38-44a0-9261-c2878dae1a24">

I think on pack we do not duplicate the `command` in the short descrption, for example

<img width="825" alt="Screenshot 2024-02-09 at 9 23 52 AM" src="https://github.com/husni-faiz/pack/assets/1181799/585f4669-b65c-4d9b-a7ef-b2a49031edfa">

I will suggest, for example with the `create` command change from 
- manifest create generates a manifest list for a multi-arch image
- Generates a manifest list for a multi-arch image
Let's remove the command from the description

jjbustamante commented on pkg/client/create_manifest.go:
```diff
@@ -2,60 +2,83 @@ package client
+	ops := parseOptsToIndexOptions(opts)
+	_, err = c.indexFactory.LoadIndex(name, ops...)
+	if err == nil {
+		return fmt.Errorf("image index with name: '%s' exists", name)
```
I will add `exits in your local storage, use 'pack manifest remove' if you want to delete it`

jjbustamante commented on pkg/client/client.go:
```diff
@@ -286,21 +309,77 @@ func (f *imageFactory) NewImage(repoName string, daemon bool, imageOS string) (i
-		opts.ManifestName,
-		opts.ManifestDir,
-		local.WithIndexMediaTypes(opts.MediaType))
+	ops = append(ops, index.WithKeychain(keychain), index.WithXDGRuntimePath(home))
```
I will suggest to keep @husni-faiz  approach here. he was doing this

```go
manifestDir := filepath.Join(packHome, "manifests")
```
basically add a root folder to organize the manifests, we also safe more information in `PackHome` and throwing the manifests there make hard for end users to understand what it is about. 

For example my local .pack home looks like

```bash
> tree ~/.pack -L 1
/Users/jbustamante/.pack
├── cnbs                    <-- this one was created after manifest create
├── completion.zsh
├── config.toml
├── download-cache
├── layout-repo
├── registry-a932275bd19c2d9e1b88fa06698fd2f5427a363d25bf87fa500691c373089381
└── registry2836348280

6 directories, 2 files
```
In pack we used to create a top folder, like `donwload-cache` or `layout-repo`

jjbustamante commented on internal/commands/manifest_create.go:
```diff
@@ -1,99 +1,73 @@
-func validateManifestCreateFlags(p *ManifestCreateFlags) error {
-	if p.Format == "" {
-		return errors.Errorf("--format flag received an empty value")
+	cmdFlags.StringVarP(&flags.format, "format", "f", "v2s2", "Format to save image index as ('OCI' or 'V2S2') (default 'v2s2')")
```
Remove `(default 'v2s2')` because it is printed duplicated after actually setting "v2s2" as default
![Screenshot 2024-02-09 at 11 06 34 AM](https://github.com/husni-faiz/pack/assets/1181799/66e9c072-1965-4210-9174-a5efb80b1497)

jjbustamante commented:
I think the Index is not created correctly, take a look at the media-type saved on disk, it should be `application/vnd.oci.image.index.v1+json` or the docker equivalent. 

![Screenshot 2024-02-09 at 11 37 34 AM](https://github.com/husni-faiz/pack/assets/1181799/9c7cb93f-0619-4d75-9aa5-f51ad71b6098)

Just to complement my last comment, the `manifest annotate` method worked correctly, but the push failed with the same error

![Screenshot 2024-02-09 at 12 05 25 PM](https://github.com/husni-faiz/pack/assets/1181799/f60613e3-34e3-4583-83e7-c177a9f4f34f)

jjbustamante commented on internal/commands/manifest_rm.go:
```diff
@@ -1,44 +1,38 @@
 		Example: `pack manifest rm cnbs/sample-package:hello-multiarch-universe \
-					cnbs/sample-package:hello-universe-windows`,
-		Long: "manifest remove will remove the specified image manifest if it is already referenced in the index",
+		cnbs/sample-package:hello-universe-windows`,
```
I think we should update the example and the documentation to include that the *digest* is required. 

I got errors like 

```bash
a digest must contain exactly one '@' separator (e.g. registry/repository@digest) saw: cnbs/sample-package:hello-universe-windows
```
And need to look for it

ep0ll replied:
now pack cli supports removing images by tags along with digests

jjbustamante commented on internal/commands/manifest_annotate.go:
```diff
@@ -1,69 +1,81 @@
+		Args:  cobra.MatchAll(cobra.ExactArgs(2), cobra.OnlyValidArgs),
+		Short: "manifest annotate modifies a manifest list (Image index) and update the platform information for an image included in the manifest list.",
+		Example: `pack manifest annotate cnbs/sample-package:hello-universe-multiarch \ 
+		cnbs/sample-package:hello-universe --arch amd64`,
```
a digest must contain exactly one '@' separator (e.g. registry/repository@digest) saw: cnbs/sample-package:hello-universe-windows

ep0ll committed:
WIP fix bugs related to imgutils

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll commented:
> I think the Index is not created correctly, take a look at the media-type saved on disk, it should be `application/vnd.oci.image.index.v1+json` or the docker equivalent.
> 
> ![Screenshot 2024-02-09 at 11 37 34 AM](https://private-user-images.githubusercontent.com/1181799/303716875-9c7cb93f-0619-4d75-9aa5-f51ad71b6098.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3Mjg2OTksIm5iZiI6MTcwNzcyODM5OSwicGF0aCI6Ii8xMTgxNzk5LzMwMzcxNjg3NS05YzdjYjkzZi0wNjE5LTRkNzUtOWFhNS1mNTFhZDcxYjYwOTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMDg1OTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTU0OWY5ZjVmY2YxYTA5OWI3NmZhY2ExODc1YzA0OWRkNTY3MDYwMTkzMDZhYWJiYmY1MmJmYzA5ODY2ZWU3OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.0TPnrNCsvBvg2MWbcpZgAHsajW7PIHMX313xkTIEV8M)
> 
> Just to complement my last comment, the `manifest annotate` method worked correctly, but the push failed with the same error
> 
> ![Screenshot 2024-02-09 at 12 05 25 PM](https://private-user-images.githubusercontent.com/1181799/303723924-f60613e3-34e3-4583-83e7-c177a9f4f34f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3Mjg2OTksIm5iZiI6MTcwNzcyODM5OSwicGF0aCI6Ii8xMTgxNzk5LzMwMzcyMzkyNC1mNjA2MTNlMy0zNGUzLTQ1ODMtODNlNy1jMTc3YTlmNGYzNGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMDg1OTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDE2ZjE4MTk5ZjMxZmFjYjAxODZmOTExNjE3ZjBhYmMxMzZiOGUxZmYyZGNlMWE1MWFkNTcyYTFhNWU0MTlmNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.-fi9mrchuVbPw9xjki-6uFVeKma-xHqoag0D6Ax3cEg)

I found the following [thing on GitHub ](https://github.com/linuxkit/linuxkit/blob/657b33805ef1acdfd198174208b48bddf7949907/docs/troubleshooting.md?plain=1#L14) even though it's not a reason for the current bug I want to take it to your notice
The above bug still exists for `manifest-full-mode` and one can push only index manifest by [ManifestHandler](https://github.com/husni-faiz/imgutil/blob/fad6445eb228f456749fdd4e220ea18974e4a94b/index.go#L92) for now and `IndexHandler` uses OCI layout format and [path#ImageIndex](https://github.com/google/go-containerregistry/blob/8dadbe76ff8c20d0e509406f04b7eade43baa6c1/pkg/v1/layout/index.go#L54)'s mediaType is defaults to OCIImageIndex even if the Index is Docker Manifest List, so when remote#Write pushing index it is setting header with `OCIImageIndex` even when pushing `Docker manifest list` which is why the error is popping up

**Any suggestions on how to deal with it?** 

1. I was thinking either to create a new oci layout for docker index which still saves index in oci layout format but when reading with [layout#ImageIndex](https://github.com/google/go-containerregistry/blob/8dadbe76ff8c20d0e509406f04b7eade43baa6c1/pkg/v1/layout/index.go#L54) will return `Docker Manifest List` format 
2. Or creating a pull request to GGCR to set header based on IndexManifest's MediaType
3. or pushing every image in the index concurrently and then push image index finally with the remote#Put to only push IndexManifest(i feel its a better option within the list)

jjbustamante commented:
> I found the following [thing on GitHub ](https://github.com/linuxkit/linuxkit/blob/657b33805ef1acdfd198174208b48bddf7949907/docs/troubleshooting.md?plain=1#L14) even though it's not a reason for the current bug I want to take it to your notice The above bug still exists for `manifest-full-mode` and one can push only index manifest by [ManifestHandler](https://github.com/husni-faiz/imgutil/blob/fad6445eb228f456749fdd4e220ea18974e4a94b/index.go#L92) for now and `IndexHandler` uses OCI layout format and [path#ImageIndex](https://github.com/google/go-containerregistry/blob/8dadbe76ff8c20d0e509406f04b7eade43baa6c1/pkg/v1/layout/index.go#L54)'s mediaType is defaults to OCIImageIndex even if the Index is Docker Manifest List, so when remote#Write pushing index it is setting header with `OCIImageIndex` even when pushing `Docker manifest list` which is why the error is popping up
> 
> **Any suggestions on how to deal with it?**
> 
> 1. I was thinking either to create a new oci layout for docker index which still saves index in oci layout format but when reading with [layout#ImageIndex](https://github.com/google/go-containerregistry/blob/8dadbe76ff8c20d0e509406f04b7eade43baa6c1/pkg/v1/layout/index.go#L54) will return `Docker Manifest List` format
> 2. Or creating a pull request to GGCR to set header based on IndexManifest's MediaType
> 3. or pushing every image in the index concurrently and then push image index finally with the remote#Put to only push IndexManifest(i feel its a better option within the list)

I saw the error you are mentioning, but that is not happening with Husni's code.  I managed to push the Image Index with Husni's pack version and the index is using docker media types

![Screenshot 2024-02-12 at 3 03 39 PM](https://github.com/husni-faiz/pack/assets/1181799/cc0f6968-9d2a-4e84-8bbe-ec89b55147bd)

ep0ll committed:
Merge branch 'test-manifest' into image-index

Signed-off-by: Sai Kiran Maggidi <107541780+WYGIN@users.noreply.github.com>

husni-faiz commented on pkg/client/docker.go:
```diff
@@ -31,4 +31,5 @@ type DockerClient interface {
 	ContainerWait(ctx context.Context, container string, condition containertypes.WaitCondition) (<-chan containertypes.WaitResponse, <-chan error)
 	ContainerAttach(ctx context.Context, container string, options types.ContainerAttachOptions) (types.HijackedResponse, error)
 	ContainerStart(ctx context.Context, container string, options types.ContainerStartOptions) error
+	ServerVersion(ctx context.Context) (types.Version, error)
```
Method duplicated in Line:24

ep0ll committed:
WIP: added couple of tests

Signed-off-by: WYGIN <wygininc@gmail.com>

husni-faiz reviewed (changes requested):
Please address the comment

ep0ll committed:
added tests for client directive

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
added tests in commands package

Signed-off-by: WYGIN <wygininc@gmail.com>

Date:      Wed Feb 28 09:46:46 2024 +0000
Changes to be committed:
	modified:   internal/commands/manifest.go
	modified:   internal/commands/manifest_add.go
	new file:   internal/commands/manifest_add_test.go
	modified:   internal/commands/manifest_annotate.go
	new file:   internal/commands/manifest_annotate_test.go
	modified:   internal/commands/manifest_create.go
	new file:   internal/commands/manifest_create_test.go
	modified:   internal/commands/manifest_exists.go
	new file:   internal/commands/manifest_exists_test.go
	modified:   internal/commands/manifest_inspect.go
	new file:   internal/commands/manifest_inspect_test.go
	modified:   internal/commands/manifest_push.go
	new file:   internal/commands/manifest_push_test.go
	modified:   internal/commands/manifest_remove.go
	new file:   internal/commands/manifest_remove_test.go
	modified:   internal/commands/manifest_rm.go
	new file:   internal/commands/manifest_rm_test.go
	new file:   internal/commands/manifest_test.go

ep0ll committed:
chore: updated imgutil to latest version

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: change features, os-features, annotations to relevant flags

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fix: bug causing ignore --features, --os-features, --urls, --annotations

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: improve code readability

Signed-off-by: WYGIN <wygininc@gmail.com>

jjbustamante committed:
Merge branch 'test-manifest' into image-index

Signed-off-by: Juan Bustamante <bustamantejj@gmail.com>

jjbustamante committed:
Merge branch 'test-manifest' into image-index

Signed-off-by: Juan Bustamante <bustamantejj@gmail.com>
